torch/utils/data/dataset.py

```diff
@@ -260,6 +265,132 @@ def __len__(self):
+    Please consult ChunkDataReader for more details.
+
+    ``ChunkDataset`` extends ``IterableDataset`` because the length of the dataset
+    (or simply the size) is unknown. Only the number of chunks is known for the dataset.
+
+    In a distributed setup, each `DataLoader` worker has an instance of ``ChunkDataset`` that uses
+    ``DistributedChunkSampler`` sampler to select chunks based on their global worker rank.
+    Once a chunk index has been selected, it is passed to ``ChunkDataReader`` that will load
+    data into the ``ChunkDataset`` internal cache. Only then ``ChunkDataset`` returns batches.
+
+    Arguments:
+        chunk_sampler (DistributedSampler or DistributedChunkSampler): Draw indices for chunks.
+            Typically used to split data amongst dataloader workers (first level of sampling)
+        chunk_reader (ChunkDataReader): Specialized reader for a given input type
+        shuffle_cache (bool): Setting `True` trigger shuffling of the internal chunk cache
+            (second level of sampling) (default: `True`)
+    """
+
+    def __init__(self, chunk_sampler, chunk_reader, shuffle_cache=True):
+        super(ChunkDataset, self).__init__()
+        assert isinstance(chunk_sampler, DistributedSampler) or isinstance(
+            chunk_sampler, DistributedChunkSampler), 'sampler must be a `DistributedSampler` or `DistributedChunkSampler`'
+        assert callable(chunk_reader), 'chunk_reader must be `callable()` and return a container with data'
+        assert isinstance(shuffle_cache, bool), 'shuffle_cache must be a `bool`'
+
+        self.chunk_sampler = chunk_sampler
+        self.chunk_reader = chunk_reader
+        self.shuffle_cache = shuffle_cache
+
+        # Internal state
+        self._chunk_sampler_iter = None
+        self._cache = []
+        self._min_cache = 1000
+        self._is_ready = False
+
+    def __iter__(self):
+        return self
+
+    def __len__(self):
+        # `IterableDataset` classes have unknown dataset size
+        raise NotImplementedError
+
+    def __next__(self):
+        r"""Returns a batch or raises exception when exhausted"""
+
+        # Run once: update global worker rank on each worker process
+        if not self._is_ready:
+            self._update_global_worker_rank()
+
+        cache_size = 0
+        if len(self._cache) > 0:
+            cache_size = len(self._cache)
+
+        if (cache_size < self._min_cache):
+            new_chunk = None
+            # Try reading a new chunk
+            # Ignore EOF here because there may be internal cache entries to be returned yet
+            try:
+                new_chunk = self.chunk_reader(next(self._chunk_sampler_iter))
+                assert isinstance(new_chunk, list), "Chunk data reader must return a `list(samples)`"
+            except StopIteration:
+                pass
+
+            # Apply second level of shuffling
+            if new_chunk and self.shuffle_cache:
+                random.shuffle(new_chunk)
```

Comment:
Shouldn't this shuffle occur across the entire cache after appending the new chunk to allow individual batches to draw from multiple chunks? As implemented it seems like chunk boundaries are preserved after shuffling in the cache, so each batch can only contain data from a single chunk except when batches happen to cross chunk boundaries.

Reply:
I think that's actually by design. The idea being that each process owns a chunk and steps through it in random order.

Reply:
Indeed, it's also part of the [paper](http://martin.zinkevich.org/publications/nips2010.pdf), see Algorithm 3.

Reply:
Doesn't each process own a fully unique sequence of chunks? Not individual chunks? If a single process only owns one chunk, there is no point to the cache at all...

When translating algorithm 3 are you assuming that the partitions with T examples are individual chunks? That does not seem like a reasonable translation as each chunk is definitely *not* a random partition of the data in most situations where something like this might be used. If we assume that a random selection of chunks is a a reasonable random partition of the data (somewhat more plausible, but still not strictly true), then the algorithm specifies that you should randomize all the data associated with the partition prior to drawing samples, not do some sort of partitioned randomization as done here.

Reply:
> Doesn't each process own a fully unique sequence of chunks? Not individual chunks? If a single process only owns one chunk, there is no point to the cache at all...
> When translating Algorithm 3 are you assuming that the partitions with T examples are individual chunks? That does not seem like a reasonable translation as each chunk is definitely not a random partition of the data in most situations where something like this might be used. If we assume that a random selection of chunks is a a reasonable random partition of the data (somewhat more plausible, but still not strictly true), then the algorithm specifies that you should randomize all the data associated with the partition prior to drawing samples, not do some sort of partitioned randomization as done here.
> If in practice this does not cause issues for all situations, would it be possible to at least add an option to do cross chunk shuffling and set the minimum cache size (i.e., effectively a randomization window in samples) in the constructor (similar to how the C++ version of this is implemented)?

@bencherian, this an idea we call mutli-chunk shuffling. Implementing such feature in this design is much easier, one probably cannot shuffle all the data per particular node due to the size but can do a reasonable shuffle between chunks. This also helps with constantly growing streams of data where new additions must be shuffled with old data to get a better randomness.